src/ListMessageSplitter.ts

```diff
@@ -0,0 +1,319 @@
+
+            // Keep looping while there are still items left.
+            // (Replaced at the end of the loop)
+            while (current_item !== undefined) {
+                // Add the new item to `sized` tentatively.
+                sized.items.push(current_item);
+
+                if (sized.size() > desired_size) {
+                    // If we went over the limit, return the last item and return the results
+                    rest.items.unshift(sized.items.pop()!)
+
+                    if (sized.items.length > 0) {
+                        return {sized, rest};
+                    } else {
+                        // If this was the first item, `sized` is empty, return null.
+                        return {sized: null, rest}
+                    }
+                }
+
+                current_item = rest.items.shift();
+            }
+
+            // ??? We somehow did not go over the size limit when testing it per item?
+            // This is a weird state, as we should have already caught this with the
+            // `this.size() <= desired_size` conditional.
+            LogService.warn("ListMessageSplitter", `encountered end of while loop, required max ${desired_size} size, current item is ${sized.size()}`)
+
+            return {sized, rest: null}
+        }
+    }
+
+    public size(): number {
+        let with_header = this.header !== null ? this.header.size() + "\n".length + "<br>".length : 0;
+
+        return with_header
+            + (this.items.length > 0 ? UL_TAG_WRAP_SIZE : 0)
+            + this.items.reduce((prev, curr, _idx, _arr) => prev + curr.size(), 0);
+    }
+
+    // Render this listing into a messagesnippet.
+    public render(): MessageSnippet {
+        let current: MessageSnippet = {
+            html: "",
+            text: ""
+        };
+
+        for (let item of this.items) {
+            current.html += `<li>${item.html}</li>`
+            current.text += ` * ${item.text}\n`
+        }
+
+        current.html = this.items.length > 0 ? `<ul>${current.html}</ul>` : "";
+
+        if (this.header !== null) {
+            current.html = this.header.html + "<br>".length + current.html;
+            current.text = this.header.text + "\n" + current.text;
+        }
+
+        return current;
+    }
+}
+
+// A class that allows splitting items and headers into multiple messages.
+export class ListMessageSplitter {
+    private items: (MessageListItem | MessageListHeader)[] = [];
+
+    constructor() {
+    }
+
+    public add_header(html: string, text: string) {
+        this.add(new MessageListHeader(html, text))
+    }
+
+    public add_paragraph(html: string, text: string) {
+        this.add(new MessageListItem(html, text))
+    }
+
+    public add(paragraph: MessageListItem | MessageListHeader) {
+        this.items.push(paragraph)
+    }
+
+    // Convert this.items into listings.
+    private get_listings(): MessageListing[] {
+        let current_listings: MessageListing[] = [];
```

Comment:
style (slightly more nitty)(here and everywhere): anything that's not being reassigned (even if the object is being mutated) should use `const`.